src/relations/database_provides.py

```diff
@@ -0,0 +1,151 @@
+
+        self.framework.observe(
+            self.charm.on[PEER].relation_changed, self._on_peer_relation_changed
+        )
+
+    # =======================
+    #  Helpers
+    # =======================
+
+    def _database_provides_relation_exists(self) -> bool:
+        database_provides_relations = self.charm.model.relations.get(DATABASE_PROVIDES_RELATION)
+        return bool(database_provides_relations)
+
+    def _get_related_app_name(self) -> str:
+        """Helper to get the name of the related `database-provides` application."""
+        if not self._database_provides_relation_exists():
+            return None
+
+        for key in self.charm.model.relations[DATABASE_PROVIDES_RELATION][0].data:
+            if type(key) == Application and key.name != self.charm.app.name:
+                return key.name
+
+        return None
+
+    # =======================
+    #  Handlers
+    # =======================
+
+    def _on_database_requested(self, event: DatabaseRequestedEvent) -> None:
+        """Handle the database requested event."""
+        if not self.charm.unit.is_leader():
+            return
+
+        self.charm.app_peer_data[MYSQL_ROUTER_PROVIDES_DATA] = json.dumps(
+            {
+                "database": event.database,
+                "extra_user_roles": event.extra_user_roles,
+            }
+        )
+
+    def _on_peer_relation_changed(self, _) -> None:
+        """Handle the peer relation changed event."""
+        if not self.charm.unit.is_leader():
+            return
+
+        if self.charm.app_peer_data.get(MYSQL_ROUTER_LEADER_BOOTSTRAPED):
+            return
+
+        if not self.charm.app_peer_data.get(MYSQL_ROUTER_REQUIRES_DATA):
+            return
+
+        database_provides_relations = self.charm.model.relations.get(DATABASE_PROVIDES_RELATION)
+        if not database_provides_relations:
+            return
+
+        parsed_database_requires_data = json.loads(
+            self.charm.app_peer_data[MYSQL_ROUTER_REQUIRES_DATA]
+        )
+        parsed_database_provides_data = json.loads(
+            self.charm.app_peer_data[MYSQL_ROUTER_PROVIDES_DATA]
+        )
+
+        db_host = parsed_database_requires_data["endpoints"].split(",")[0].split(":")[0]
+        mysqlrouter_username = parsed_database_requires_data["username"]
+        mysqlrouter_user_password = self.charm._get_secret("app", "database_password")
+
+        try:
+            MySQLRouter.bootstrap_and_start_mysql_router(
+                mysqlrouter_username,
+                mysqlrouter_user_password,
+                self._get_related_app_name(),
+                db_host,
+                "3306",
+            )
+        except MySQLRouterBootstrapError:
+            self.charm.unit.status = BlockedStatus("Failed to bootstrap mysqlrouter")
+            return
+
+        provides_relation_id = database_provides_relations[0].id
+        application_username = f"application-user-{provides_relation_id}"
```

Comment:
Shouldn't this follow the following [pattern](https://github.com/canonical/mysql-operator/blob/8d087a7abf808c701e369972c752c7592c5b3a17/lib/charms/mysql/v0/mysql.py#L450) that is used by other data platforms charms? (relation-{provides_relation_id}) Maybe you can ignore this, as this charm is supposed to be related to only one application.

Reply:
I am using `application-user-{relation_id}` so that the application usernames do not collide with the database usernames (where the relation_id could be the same)

Reply:
It raises the question on how the uniqueness of relation-id is handled by juju, i.e. is it scoped on app, model or controller?

Reply:
i believe it is scoped to each application. i saw the `id=3` being used for both the shared-db and database relation interfaces